plugins/amqp_trace_plugin/include/eosio/amqp_trace_plugin/amqp_handler.hpp

```diff
@@ -0,0 +1,83 @@
+#pragma once
+#include <amqpcpp.h>
+#include <amqpcpp/libboostasio.h>
+#include <amqpcpp/linux_tcp.h>
+
+#include <fc/log/logger.hpp>
+
+namespace eosio {
+
+class amqp_handler : public AMQP::LibBoostAsioHandler {
+public:
+   explicit amqp_handler( boost::asio::io_service& io_service ) : AMQP::LibBoostAsioHandler( io_service ) {}
+
+   void onError( AMQP::TcpConnection* connection, const char* message ) override {
+      elog( "amqp connection failed: ${m}", ("m", message) );
+   }
+
+   uint16_t onNegotiate( AMQP::TcpConnection* connection, uint16_t interval ) override {
+      return 0; // disable heartbeats
+   }
+};
+
+/// Designed to work with io_service running on a dedicated thread.
+class amqp {
+public:
+   using on_error_t = std::function<void(const std::string& err)>;
+   amqp( boost::asio::io_service& io_service, const std::string& address, std::string name, on_error_t on_err )
+         : name_( std::move( name ) )
+         , on_error( std::move(on_err) )
+   {
+      AMQP::Address amqp_address( address );
+      ilog( "Connecting to AMQP address ${a} - Queue: ${q}...", ("a", std::string( amqp_address ))( "q", name_ ) );
+
+      handler_ = std::make_unique<amqp_handler>( io_service );
+      connection_ = std::make_unique<AMQP::TcpConnection>( handler_.get(), amqp_address );
+      channel_ = std::make_unique<AMQP::TcpChannel>( connection_.get() );
+      declare_queue();
+   }
+
+   void publish( const std::string& exchange, const std::string& correlation_id, const char* data, size_t data_size ) {
+      AMQP::Envelope env( data, data_size );
+      env.setCorrelationID( correlation_id );
+      channel_->publish( exchange, name_, env, 0 );
+   }
+
+   auto& consume() { return channel_->consume( name_ ); }
+
+   void ack( uint64_t delivery_tag ) {
+      channel_->ack( delivery_tag );
+   }
+
+   void reject( uint64_t delivery_tag ) {
+      channel_->reject( delivery_tag );
+   }
+
+private:
+   void declare_queue() {
+      std::promise<void> connected;
+      auto& queue = channel_->declareQueue( name_, AMQP::durable );
```

Comment:
`declare_queue` is executing in the application thread, right?  I'm not sure that this is safe.  The `io_context` is already running, so the operation may complete at any time, and has a race with setting the handler.  It looks to me as though accessing a `Channel` from multiple threads is unsafe, in general.  (I wasn't really sure from the amqp-cpp documentation, so I poked around the implementation a bit).

Reply:
Yeah I'm surprised there is nothing explicit in the docs about threading, but I've seen some comments from Copernica devs that indicate as such
https://github.com/CopernicaMarketingSoftware/AMQP-CPP/issues/238#issuecomment-525279268

Reply:
Changed to call `declare_queue` on amqp thread.